kqueue.c

```diff
@@ -211,65 +228,32 @@ kq_build_changes_list(const struct event_changelist *changelist,
 	for (i = 0; i < changelist->n_changes; ++i) {
 		struct event_change *in_ch = &changelist->changes[i];
 		struct kevent *out_ch;
-		if (n_changes >= kqop->changes_size - 1) {
-			int newsize;
-			struct kevent *newchanges;
-
-			if (kqop->changes_size > INT_MAX / 2 ||
-			    (size_t)kqop->changes_size * 2 > EV_SIZE_MAX /
-			    sizeof(struct kevent)) {
-				event_warnx("%s: int overflow", __func__);
-				return (-1);
-			}
-
-			newsize = kqop->changes_size * 2;
-			newchanges = mm_realloc(kqop->changes,
-			    newsize * sizeof(struct kevent));
-			if (newchanges == NULL) {
-				event_warn("%s: realloc", __func__);
-				return (-1);
-			}
-			kqop->changes = newchanges;
-			kqop->changes_size = newsize;
-		}
 		if (in_ch->read_change) {
-			out_ch = &kqop->changes[n_changes++];
+			out_ch = &kqop->events[n_changes++];
 			kq_setup_kevent(out_ch, in_ch->fd, EVFILT_READ,
 			    in_ch->read_change);
+			if (n_changes == kqop->events_size && kq_grow_events(kqop)) {
```

Comment:
Should we check for growing before access, i.e. before `out_ch` assignment?

Reply:
That would be needless. n_changes starts with zero and its value increment by one after one kevent in events is used (for every read or write event). That increased value will be checked each time we consume a kevent, which ensures that the size of events is doubled right after the old size is used up and makes sure there are always rooms for the next event and subsequent events.